### ISSUE TITLE
@ExpectMaxQueryExecutionTime alert message always displays "0 ms" as max

### ISSUE THREAD

**Issue by FTarfasse** [quick-perf/quickperf#93]
Hello Jean,

I noticed the following, I'd be happy to have your opinion on this: 

**Describe the bug**
When using [@ExpectMaxQueryExecutionTime](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/annotation/ExpectMaxQueryExecutionTime.java), the console alert message always displays 0 ms in "_**[PERF] At least one request exceeds the max expected query execution time <0 ms>.**_" and does not remind the value and unit set on the annotation.
Example :
`@ExpectMaxQueryExecutionTime(value = 2, unit = TimeUnit.NANOSECONDS)`

**Expected behavior**
"_**[PERF] At least one request exceeds the max expected query execution time < 2 ns >.**_" where value and unit are the ones set in the annotation (see example above).

**Actual behavior**
"[PERF] At least one request exceeds the max expected query execution time <0 ms>."

**To Reproduce**
Here is [an example](https://github.com/FTarfasse/quickperf/commit/1adfe4c0cb78dfb90c7afdb0b5734ebca0dc27a9) using the @ExpectMaxInsertTest. 
**CAREFUL !** You should ONLY launch the AClassHavingAMethodAnnotatedWithExpectMaxInsert test (first of the three tests in the class), it should fail. When launching all the tests on the class, the test passes.
I used my IDE in order to launch only the first test (Intellij).

**Versions**
- QuickPerf: v1.0
- JDK: openjdk version "11.0.7"
- OS: Linux / Ubuntu 18.04
- Database (if SQL annotation bug): JPA

**Additional context** (Add any other context about the problem here.)
From my understanding, the issue is linked to the corresponding [verifier class](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/time/SqlQueryMaxExecutionTimeVerifier.java). Other verifier classes (such as the one in the core [time package](https://github.com/quick-perf/quickperf/blob/master/core/src/main/java/org/quickperf/time/ExecutionTimeFormatter.java)) often use a dedicated formatter class to help format and display this kind of informations. Time package is restricted to core annotations and the [sql package formatter](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/formatter/QuickPerfSqlFormatter.java) doesn't (and is not supposed to) format execution time.

Kind regards,
Fabrice

**Comment by jeanbisutti** [comment]
Hello Fabrice,

Thanks a lot for this report.

Indeed, there is an issue in the error report when the threshold is expressed with `TimeUnit.MICROSECONDS` or `TimeUnit.NANOSECONDS`.

In `SqlQueryMaxExecutionTimeVerifier` class:

```java
    @Override
	public PerfIssue verifyPerfIssue(ExpectMaxQueryExecutionTime annotation, ExecutionTime measure) {
		
		ExecutionTime maxExpectedSqlExecutionTime = new ExecutionTime(annotation.value(), annotation.unit());
		
		if(measure.isGreaterThan(maxExpectedSqlExecutionTime)) {

			String message = "At least one request exceeds the max expected query execution time <" + maxExpectedSqlExecutionTime + ">.";
			
```
If we look at the `toString()` implementation of `ExecutionTime`:
```java
	@Override
	public String toString() {
		return this.unit.toMillis(value) + " ms";
	}
```

If `unit` value is `TimeUnit.MICROSECONDS` or `TimeUnit.NANOSECONDS` then ` unit.toMillis(value)` value can be zero.

QuickPerf uses [ttddyy](https://github.com/ttddyy/datasource-proxy) to intercept the SQL queries. In ttddy, the execution time is given in milliseconds.

In QuickPerf, I think that a threshold less than to the millisecond does not make sense.

I would prefer to have an API like this:

```
public @interface ExpectMaxQueryExecutionTime {

	int thresholdInMilliSeconds();
	
```
	
I will be glad to have your opinion.

You are very welcome if you would like to work on developing this new API.

I would also like to remove `value` and `unit` annotation elements. QuickPerf is in release candidate version (hoping not for a long time), so it's still the moment to do it.

 Kind regards,

Jean

**Comment by FTarfasse** [comment]
Hi Jean,

Thanks again for the clarifications (especially on ttddyy).

Your idea of removing the time unit from the interface is appealing (always having an int value representing milliseconds). I can make a draft to see where this will lead me, I'm sure I will ask you more questions later :).

I will give you an update soon.

Kind regards,
Fabrice

**Comment by FTarfasse** [comment]
Hi Jean,

Sorry for the delay, I made a draft starting from your idea (don't consider the two previous commits, I forgot Git reports everything even when cautiously working on a separate branch).
The "good" commit is [here](https://github.com/FTarfasse/quickperf/commit/b86021f2a436a93a2bdbfc6071fa5746b3f4ec56).

Notice I tweaked the alert message from [SqlQueryMaxExecutionTimeVerifier](https://github.com/quick-perf/quickperf/blob/master/sql/sql-annotations/src/main/java/org/quickperf/sql/time/SqlQueryMaxExecutionTimeVerifier.java).
When testing and setting the annotation on several tests I was confused as I had no idea of the query execution time that happened effectively (plus the effective execution time can vary from time to time). I played for a time with the @MeasureExecutionTime annotation until I realised it wasn't relevant.

So, even though the [SQL Executions] message displays the execution time for each query, I modified the alert to display both reality and expectation as following :

```
[PERF] At least one request exceeds the max expected query execution time:
    Effective execution time <1 ms>
    Expected execution time <0 ms>
```

I'd be happy to have your feedback.

P.S. : there is still no test that tests the alert message itself (which is I feel the origin of the issue in a way), I simply updated the SqlQueryMaxExecutionTimeVerifier). Let me know if you want me to set a new test class.

Kind regards,
Fabrice

**Comment by jeanbisutti** [comment]
Hi Fabrice,

Many thanks for your work. Don't worry about the delay. You are welcome.

> So, even though the [SQL Executions] message displays the execution time for each query, I modified the alert to display both reality and expectation as following 

Good idea Fabrice.

"Effective" may seem ambiguous. 

Perhaps, we could propose something looking like 

```
[PERF] Query execution time expected to be less than  <1 ms>.  
       At least one query has a greater execution time. The greater query execution time is <2ms>.
```

With several queries, the user can find the query taking a long time by looking at the console:

```
[SQL EXECUTIONS]
	Time:1, Success:True, Type:Prepared, Batch:False, QuerySize:1, BatchSize:0, Query:["
    select
        book0_.id as id1_0_,
        book0_.isbn as isbn2_0_,
        book0_.title as title3_0_ 
    from
        Book book0_"], Params:[()]
```

`Time:1 ` means that the query takes 1 ms. We could update the documentation.

> P.S. : there is still no test that tests the alert message itself (which is I feel the origin of the issue in a way), I simply updated the SqlQueryMaxExecutionTimeVerifier). Let me know if you want me to set a new test class.

You are right. The message content is not automatically checked. [I put a comment in your Github repo](https://github.com/FTarfasse/quickperf/commit/b86021f2a436a93a2bdbfc6071fa5746b3f4ec56#diff-7fc174f6570f56d5e49dd253af0ee087L39).

Thanks a lot for your relevant remarks Fabrice.

Kind regards,

Jean

**Comment by FTarfasse** [comment]
Hi Jean,

Happy to help ! Here is an updated [commit](https://github.com/quick-perf/quickperf/commit/fe9d338abae42be94fd42566e1ff10f31fc37172) based on your feedbacks. To sum up:
- added assertj assertions as discussed,
- refactored (presentation) to match existing BDD style of other tests,
- added assertion testing for the message itself.
If ok I'll rebase and squash to do a proper PR on the master branch. I'll update the documentation and push to the doc repo in parallel.

Kind regards,
Fabrice

**Comment by jeanbisutti** [comment]
Hi Fabrice,

It's perfect!

I will update the wiki from your proposal just after the next release, together with quickperf examples.

Many thanks for your contributions Fabrice!

Kind regards,

Jean

**Comment by FTarfasse** [comment]
It's been a pleasure !

**Comment by jeanbisutti** [comment]
Fixed with bbb561bd1eeeeab228e12688b14a571bfd9592b9
